src/types/connection/connection_type.rs

```diff
@@ -1,97 +1,136 @@
-use crate::types::connection::cursor::Cursor;
-use crate::types::connection::edge::Edge;
-use crate::types::connection::page_info::PageInfo;
+use crate::connection::edge::Edge;
+use crate::connection::page_info::PageInfo;
+use crate::types::connection::{CursorType, EmptyEdgeFields};
 use crate::{
-    do_resolve, registry, Context, ContextSelectionSet, EmptyEdgeFields, Error, ObjectType,
+    do_resolve, registry, Context, ContextSelectionSet, Error, FieldResult, ObjectType,
     OutputValueType, Positioned, QueryError, Result, Type,
 };
 use async_graphql_parser::query::Field;
+use futures::{Stream, TryStreamExt};
 use indexmap::map::IndexMap;
 use inflector::Inflector;
 use itertools::Itertools;
 use std::borrow::Cow;
 
+/// The record type output by the data source
+pub struct Record<C, E, T> {
+    cursor: C,
+    edge: E,
+    element: T,
+}
+
+impl<C, E, T> Record<C, E, T> {
+    pub fn new(cursor: C, edge: E, element: T) -> Self {
+        Self {
+            cursor,
+            edge,
+            element,
+        }
+    }
+}
+
+impl<C, T> Record<C, EmptyEdgeFields, T> {
+    pub fn new_without_edge_fields(cursor: C, element: T) -> Self {
+        Self {
+            cursor,
+            edge: EmptyEdgeFields,
+            element,
+        }
+    }
+}
+
 /// Connection type
 ///
-/// Connection is the result of a query for `DataSource`,
-/// If the `T` type is `OutputValueType`, you can return the value as a field function directly,
-/// otherwise you can use the `Connection::map` function to convert to a type that implements `OutputValueType`.
-/// `E` is an extension object type that extends the edge fields.
+/// Connection is the result of a query for `DataSource`.
 pub struct Connection<T, E: ObjectType + Send = EmptyEdgeFields> {
     /// The total number of records.
-    pub total_count: Option<usize>,
+    pub(crate) total_count: Option<usize>,
 
     /// Information about pagination in a connection.
-    pub page_info: PageInfo,
+    pub(crate) page_info: PageInfo,
 
     /// All records of the current page.
-    pub nodes: Vec<(Cursor, E, T)>,
+    pub(crate) nodes: Vec<Record<String, E, T>>,
 }
 
-impl<T, E: ObjectType + Send> Connection<T, E> {
-    /// Create a connection object.
-    pub fn new(
-        total_count: Option<usize>,
+impl<T: OutputValueType + Send, E: ObjectType + Send> Connection<T, E> {
+    pub fn empty() -> Self {
```

Comment:
Givent the empty values, I guess `Connection` and `PageInfo` could derive `Default`?

Reply:
I tried it and failed to compile it. 🙁

Reply:
Strange, looks like the compiler want C, T and E to implement default as well, even if they're only used in a Vec and Option, which impl Default whatever the type :/